mlir/lib/Mitigation/Transforms/MitigationMethods/Zne.cpp

```diff
@@ -34,33 +35,142 @@
 namespace catalyst {
 namespace mitigation {
 
+bool containsQnodes(func::FuncOp funcOp)
+{
+    bool containsQnodes = false;
+    funcOp.walk([&](func::CallOp op) {
+        auto insideFuncOp =
+            SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(op, op.getCalleeAttr());
+        if (insideFuncOp->hasAttr("qnode")) {
+            containsQnodes = true;
+            return WalkResult::interrupt();
+        }
+        return WalkResult::advance();
+    });
+    return containsQnodes;
+}
+
+func::FuncOp createZneFunc(func::FuncOp funcOp, PatternRewriter &rewriter)
+{
+    auto loc = funcOp.getLoc();
+    TypeRange originalTypes = funcOp.getArgumentTypes();
+    SmallVector<Type> typesFolded(originalTypes.begin(), originalTypes.end());
+    Type indexType = rewriter.getIndexType();
+    typesFolded.push_back(indexType);
+    FunctionType fnFoldedType = FunctionType::get(funcOp.getContext(),
+                                                  /*inputs=*/typesFolded,
+                                                  /*outputs=*/funcOp.getResultTypes());
+    std::string fnFoldedName = funcOp.getName().str() + ".zne";
+    auto fnFoldedOp = rewriter.create<func::FuncOp>(loc, fnFoldedName, fnFoldedType);
+
+    rewriter.cloneRegionBefore(funcOp.getBody(), fnFoldedOp.getBody(), fnFoldedOp.end());
+
+    Block *fnFoldedOpBlock = &fnFoldedOp.getBody().front();
+    rewriter.setInsertionPointToStart(fnFoldedOpBlock);
```

Comment:
This set one also seems out of place since no new ops are created after it (in this scope). It would be safer to set the IP where the ops are created (better locality).

Reply:
This one was too much, removed